### ISSUE TITLE
When Upload Speed is adjusted, the download speed drops to 0kb/s

### ISSUE THREAD

**Issue by jdsys271** [Tribler/tribler#5290]
##### Tribler version/branch+revision: #####
V7.5.0-rc4

##### Operating system and version: #####
Xubuntu 18.04

##### Steps to reproduce the behavior: #####
- Download torrent
- Set Upload Speed to 5
- Watch how the upload speed **INCLUDING the download speed** drops from 2mb/s to 0b/s

##### Expected behavior: #####
If i set the Upload speed to 5kb/s or 1kb/s, it must set the Upload speed, not the download speed as well.

##### Actual behavior: #####
Whatever you set on Upload Speed, they both drop to 0kb/s within minutes during a download.

Im testing out Tribler and Im not sure why this is happening?
Is this because you dont want Leachers to be unfair towards the torrent community? By forcing them to use an unlimited upload rate so that they can have an unlimited download rate?

If this is not a bug then this is really sneaky and unfair.  Some users who live in countries with expensive CAPPED internet means that being an uploader obviously consumes their bandwidth.

If this is not a bug, then can you explain if Tribler's mission is to FORCE users to upload?
What happens if they dont want to upload? Tribler kills their download speed?
Please explain.

**Comment by ichorid** [comment]
Thank you for your report!
I believe what you witnessed when you cap the upload rate to such a low limit is not a bug, but a _feature_ of the BitTorrent protocol called "[choking](https://wiki.theory.org/index.php/BitTorrentSpecification#Choking_and_Optimistic_Unchoking)".

From BitTorrent wiki:
>If they become interested, the downloader with the worst upload rate gets choked. If a client has a complete file, it uses its upload rate rather than its download rate to decide which peers to unchoke. 

When you set your upload rate to such a small value, you're essentially showing other peers that you're not going to reciprocate their sharing files with you. In fact, it is BitTorrent's mission to FORCE users to upload, and that is the sole reason why it succeeded.

**Comment by ichorid** [comment]
So, just to be sure, I ran the libtorrent's `client_test` utility with rate limit of 5k, using Fedora Games image `magnet:?xt=urn:btih:a573d0ebbe7c7a87f4d8794a7432392cde3b8189\&dn=Fedora-Games-Live-i386-30`

```bash
client_test --upload_rate_limit=5000 "magnet:?xt=urn:btih:a573d0ebbe7c7a87f4d8794a7432392cde3b8189"
```

The torrent currently is indicated to have 33 seeds and 3 leeches, which means that it is not overseeded, and choking still matters. As expected, the download rate would instantly spike to hundreds of Kb/s, then drop to 5-10 Kb/s. On average, it stays on 5-50 Kb/s.
Note that on overseeded torrents choking almost never kicks in, so upload rate does not affect the download performance.

**Comment by jdsys271** [comment]
> Note that on overseeded torrents choking almost never kicks in, so upload rate does not affect the download performance.

well i tested downloading [this](magnet://showrss.info?xt=urn:btih:19584CF6C2A6026D253269FB8F1AB927F5663625&dn=Westworld+S03E05+WEB+H264+XLF&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969%2Fannounce&tr=udp%3A%2F%2Ftracker.leechers-paradise.org%3A6969%2Fannounce&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337%2Fannounce&tr=http%3A%2F%2Ftracker.trackerfix.com%3A80%2Fannounce) for example (which is overseeded). As it was downloading with over 3mb/s i changed the Upload setting from 0 to 5kb/s and it immediately choked running down to 0kb/s, then i change the setting again to 0kb/s and the Download speed shot up again.

**Comment by ichorid** [comment]
Could you please only post links to _official_ content on out GitHub pages? Thank you :wink:

**Comment by ichorid** [comment]
AFAIK libtorrent counts "technical" packets like utp ACKs as a part of the upload traffic. This means that if you set your upload traffic rate limit to some really low value, your client will not be able to send "acknowledge" packets back to others fast enough to keep up with them sending data to you. From their standpoint it will look like your connection is congested, so they will send you data slower.

I tried an overseeded Ubuntu torrent with 5 kb/s limit, it gave me about 100kb/s download rate. With 10 kb/s upload limit I got 200-500 kb/s download rate. Even a prehistoric modem :sauropod: is able to sustain more than 50kb/s upload rate. With that rate, I got 10-100 Mbit/s downloads.
